src/ast/passes/unstable_feature.cpp

```diff
@@ -0,0 +1,107 @@
+#include <cstring>
+#include <format>
+#include <string>
+
+#include "ast/ast.h"
+#include "ast/passes/unstable_feature.h"
+#include "ast/visitor.h"
+#include "bpftrace.h"
+#include "config.h"
+
+namespace bpftrace::ast {
+
+namespace {
+
+std::string get_warning(const std::string &feature)
+{
+  return std::format("Script is using an unstable feature. To prevent this "
+                     "warning you must explicitly enable the unstable "
+                     "feature in the config e.g. {}=1",
+                     feature);
+}
+
+std::string get_error(std::string &&feature)
+{
+  return std::format("Feature not enabled by default. To enable "
+                     "this unstable feature, set the config flag to 1. {}=1",
+                     feature);
+}
+
+class UnstableFeature : public Visitor<UnstableFeature> {
+public:
+  explicit UnstableFeature(BPFtrace &bpftrace) : bpftrace_(bpftrace) {};
+
+  using Visitor<UnstableFeature>::visit;
+  void visit(AssignConfigVarStatement &assignment);
+  void visit(MapDeclStatement &decl);
+  void visit(Import &imp);
+  void visit(Macro &macro);
+
+private:
+  BPFtrace &bpftrace_;
+  std::unordered_set<std::string> enabled_unstable_features;
+};
+
+} // namespace
+
+void UnstableFeature::visit(AssignConfigVarStatement &assignment)
+{
+  // If set explicitly, we don't need to warn of use
+  if (bpftrace_.config_->is_unstable(assignment.var)) {
```

Comment:
Can we just explicitly make this configuration variable a tri-state? (false/warn/true) And then just refer to the config here in this analyser, rather than introducing a new separate map that basically tracks whether it is explicitly set or not.

Reply:
I made it a tri-state but I still need the map to avoid issuing the same warning multiple times.

Reply:
Yeah SG. We could also mutate the config value on first warning, but is undesirable for other reasons. This seems reasonable to me.